doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+Panda3D C++ Style Guide
+=======================
+
+Rather than try to justify every decision made, this document attempts to be as
+at-a-glance as possible.  Examples provided where necessary to illustrate best
+practices.
+
+Note that this style guide is, for the most part, not very strict, and there
+are exceptions to every rule.  However, this guide does serve to illustrate the
+preferred style for the entire codebase.
+
+Naming conventions
+------------------
+
+- **Classes**: `CamelCaseFirstUpper`
+- **Filenames**: `lowerCamelCase.h`, `lowerCamelCase.cxx`
+  - Note, every header file must have a unique filename.  They are all
+    installed into the same path.
+- **Functions**: `lower_snake_case_without_prefix`
+- **Local variables**: `lower_snake_case_without_prefix`
+- **Member variables**: `_lower_snake_case_with_initial_score`
+
+Indentation
+-----------
+
+Always use two spaces to indent.  Tabs should never appear in a C++ file.
+
+Access modifiers (`public`/`private`/`protected`/`PUBLISHED`) are never
+indented, and do not affect the indentation of the surrounding lines.
+
+Spaces
+------
+
+Always put one space after a control keyword, before the condition's
+parentheses.  Don't pad the inside of parentheses with spaces unless there are
+several nested parenthetical expressions and the extra spaces contribute
+substantially to readability.
+
+    if (x == y) {
+
+Don't place a space after a function name (declarations, definitions, or calls).
+
+    void function_name();
+    ...
+    function_name();
+
+Always place a space after each comma in an arguments list.
+
+    function_call(x, y, z);
```

Comment:
What about default argument assignments?  Do we want to follow the Python convention of using no spaces around `=` for default arguments, or be consistent with other uses of the `=` keyword and put spaces around it? eg.
```cpp
// Python convention
void function_name(double x=3.0);
// Current convention / consistency
void function_name(double x = 3.0);```

Reply:
From a quick look, it seems that the current code mostly uses `double x=3.0`, maybe we should then stick with that?

Reply:
I think all of David's code uses spaces around `=`, except for the `=0` behind virtual functions.

Reply:
I don't have a strong opinion on that, we can settle for whatever you prefer

Reply:
Maybe we should consider [these recommendations](https://google.github.io/styleguide/cppguide.html#Functions) in general and [these](https://google.github.io/styleguide/cppguide.html#Default_Arguments) on default arguments precisely.